runtime/common/src/claims.rs

```diff
@@ -282,17 +382,45 @@ impl<T: Trait> Module<T> {
 		v.extend(rev.into_iter().rev());
 		v.extend_from_slice(&prefix[..]);
 		v.extend_from_slice(what);
+		v.extend_from_slice(extra);
 		v
 	}
 
 	// Attempts to recover the Ethereum address from a message signature signed by using
 	// the Ethereum RPC's `personal_sign` and `eth_sign`.
-	fn eth_recover(s: &EcdsaSignature, what: &[u8]) -> Option<EthereumAddress> {
-		let msg = keccak_256(&Self::ethereum_signable_message(what));
+	fn eth_recover(s: &EcdsaSignature, what: &[u8], extra: &[u8]) -> Option<EthereumAddress> {
+		let msg = keccak_256(&Self::ethereum_signable_message(what, extra));
 		let mut res = EthereumAddress::default();
 		res.0.copy_from_slice(&keccak_256(&secp256k1_ecdsa_recover(&s.0, &msg).ok()?[..])[12..]);
 		Some(res)
 	}
+
+	fn process_claim(signer: EthereumAddress, dest: T::AccountId) -> DispatchResult {
+		let balance_due = <Claims<T>>::get(&signer)
+			.ok_or(Error::<T>::SignerHasNoClaim)?;
+
+		let new_total = Self::total().checked_sub(&balance_due).ok_or(Error::<T>::PotUnderflow)?;
+
+		// Check if this claim should have a vesting schedule.
+		if let Some(vs) = <Vesting<T>>::get(&signer) {
+			// If this fails, destination account already has a vesting schedule
+			// applied to it, and this claim should not be processed.
+			T::VestingSchedule::add_vesting_schedule(&dest, vs.0, vs.1, vs.2)
+				.map_err(|_| Error::<T>::DestinationVesting)?;
+		}
+
+		CurrencyOf::<T>::deposit_creating(&dest, balance_due);
+		<Total<T>>::put(new_total);
+		<Claims<T>>::remove(&signer);
+		<Vesting<T>>::remove(&signer);
+		Signing::remove(&signer);
+		Preclaims::<T>::remove(&dest);
```

Comment:
This is potentially nasty?

A user with a preclaim could call `claim` and their preclaim would be wiped away

Solution here would be to verify that `dest` account in claim does not have a preclaim

Reply:
fixed.